ibis/pandas/udf.py

```diff
@@ -0,0 +1,245 @@
+        return scalar_types
+
+    @rule_to_python_type.register(dt.Array)
+    def array_rule(rule, groupby=False):
+        return (list,)
+
+    @rule_to_python_type.register(dt.Map)
+    def map_rule(rule, groupby=False):
+        return (dict,)
+
+    @rule_to_python_type.register(dt.Struct)
+    def struct_rule(rule, groupby=False):
+        return (collections.OrderedDict,)
+
+    @rule_to_python_type.register(dt.String)
+    def string_rule(rule, groupby=False):
+        return six.string_types
+
+    @rule_to_python_type.register(dt.Integer)
+    def int_rule(rule, groupby=False):
+        return six.integer_types + (np.integer,)
+
+    @rule_to_python_type.register(dt.Floating)
+    def float_rule(rule, groupby=False):
+        return (float, np.floating)
+
+
+def nullable(datatype):
+    """Return the signature of a scalar value that is allowed to be NULL (in
+    SQL parlance).
+
+    Parameters
+    ----------
+    datatype : ibis.expr.datatypes.DataType
+
+    Returns
+    -------
+    Tuple[Type]
+    """
+    return (type(None),) if datatype.nullable else ()
+
+
+def signature(input_type, klass):
+    """Compute the appropriate signature for an ibis Node from a list of input
+    types.
+
+    Parameters
+    ----------
+    input_type : List[ibis.expr.datatypes.DataType]
+        A list of ibis DataType instances
+    klass : Type
+        pd.Series or SeriesGroupBy
+
+    Returns
+    -------
+    Tuple[Type]
+        A tuple of types appropriate for use in a multiple dispatch signature.
+
+    Examples
+    --------
+    >>> import pandas as pd
+    >>> from pandas.core.groupby import SeriesGroupBy
+    >>> import ibis.expr.datatypes as dt
+    >>> input_type = [dt.int64, dt.double]
+    >>> signature(input_type, pd.Series)
+    ((int64, <class 'NoneType'>), (double, <class 'NoneType'>))
+    >>> input_type = [dt.Int64(nullable=False), dt.Double(nullable=False)]
+    >>> signature(input_type, SeriesGroupBy)
+    (int64, double)
+    """
+    return (
+        (klass,) + rule_to_python_type(r) + nullable(r) for r in input_type
+    )
+
+
+def udf(input_type, output_type):
+    """Define a UDF (user-defined function) that operates element wise on a
+    Pandas Series.
+
+    Parameters
+    ----------
+    input_type : List[ibis.rules]
+        A rules encoding the abstract type of each argument. These are found in
+        :mod:`~ibis.expr.rules`
+    output_type : ibis.rules or ibis.expr.datatypes.DataType
+        The return type of the function.
+
+    Examples
+    --------
+    >>> import ibis
+    >>> @udf(input_type=[dt.string], output_type=dt.int64)
+    ... def my_string_length(series, **kwargs):
+    ...     return series.str.len() * 2
+    """
+    def wrapper(func):
+        # generate a new custom node
+
+        UDFNode = type(
+            func.__name__ + '_udf_node',
+            (ops.ValueOp,),
+            dict(input_type=input_type, output_type=output_type.array_type)
```

Comment:
Should not an `udf` work on mixed scalar and array inputs/outputs too? 
If so the `rules.shape_like_arg[s]` might be required.